04-onion-routing.md

```diff
@@ -786,12 +1028,26 @@ The association between the forward and return packets is handled outside of
 this onion routing protocol, e.g. via association with an HTLC in a payment
 channel.
 
+Error handling for HTLCs with `blinding_point` is particularly fraught,
+since differences in implementations (or versions) may be leveraged to
+de-anonymize elements of the blinded path. Thus the decision to allow
+only errors from the final node to propagate, and turn every other error
+into `invalid_onion_blinding` which will be converted to a normal onion
+error by the introduction point.
+
 ### Requirements
 
 The _erring node_:
   - SHOULD set `pad` such that the `failure_len` plus `pad_len` is equal to 256.
     - Note: this value is 118 bytes longer than the longest currently-defined
     message.
+  - If `blinding_point` is set in the incoming `update_add_htlc`:
+    - if it is not the final node (including if the `onion_routing_packet` is malformed):
+	    - SHOULD forward normal downstream `update_fail_htlc` errors (which the final node may send).
+      - MUST return `invalid_onion_blinding` for any local error or other downstream errors.
+    - otherwise (it is the final node):
+      - MAY return normal errors that help senders retry the payment (e.g. `mpp_timeout`).
```

Comment:
Isn't the blinding causing a problem for the sender to decrypt a normal error with the shared secret?

Reply:
No it doesn't, the blinding doesn't have any impact here, you just use shared secrets to generate errors. However, https://github.com/lightning/bolts/pull/765/commits/eed2ab0c30ad7f93e3b2641ca9d7ade32f3d121d is making errors stricter, which makes this whole point moot.

Reply:
Yes, with that it doesn't matter anymore.

But still interested to understand how it was previously intended to work. As far as I understand, a blinded node now has two shared secrets? One with the receiver for the encrypted routing instructions, and another one with the sender so that failures can be returned?

Reply:
There is just one shared secret, exactly like the non-blinded case. The only difference is the _private key_ that intermediate nodes need to use to derive that shared secret: in normal payments they directly use the private key associated with their `node_id`, in blinded payments they have to tweak that private key. But the shared secrets mechanism is left unchanged.

Reply:
Ok, clear! Would have been useful to add that to the docs in case other readers wonder the same thing. But not necessary anymore as you deleted it all now :)

Reply:
Hmm but if there is just one shared secret that is used by a node in the blinded path to:

* decrypted the blinded node onion payload coming from the receiver, _and_
* encrypt a potential failure message to the sender

Isn't the sender then also able to decrypt the node onion payload?

Reply:
I don't see how...?

Reply:
If the sender is able to decrypt a failure message, it must have access to the same secret that the blinded node uses to decrypt the onion payload? (Assuming there is only one shared secret)

Reply:
I think I now understand what you mean by "two shared secrets". There are indeed two completely independent shared secrets.

When the **recipient** creates the encrypted payloads for nodes inside the blinded route, it uses ECDH with those nodes, which yields what we could call the _recipient_ shared secrets.

When the **sender** creates onion payloads for every node in the route (blinded or not blinded), it also uses ECDH with those nodes, which yields what we could call the _sender_ shared secrets.

Those are completely independent and the sender doesn't have access to anything related to the recipient shared secrets.